target_snowflake/db_sync.py

```diff
@@ -403,9 +403,35 @@ def delete_from_stage(self, stream, s3_key):
         self.upload_client.delete_object(stream, s3_key)
 
     def copy_to_archive(self, s3_source_key, s3_archive_key, s3_archive_metadata):
-        """Copy file from snowflake stage to archive"""
-        self.logger.info('Copying %s to archive location %s', s3_source_key, s3_archive_key)
-        self.upload_client.copy_object(s3_source_key, s3_archive_key, s3_archive_metadata)
+        """
+        Copy file from snowflake stage to archive.
+
+        s3_source_key: The s3 key to copy, assumed to exist in the bucket configured as 's3_bucket'
+
+        s3_archive_key: The key to use in archive destination. This will be prefixed with the config value
+                        'archive_load_files_s3_prefix'. If none is specified, 'archive' will be used as the prefix.
+
+                        As destination bucket, the config value 'archive_load_files_s3_bucket' will be used. If none is
+                        specified, the bucket configured as 's3_bucket' will be used.
+
+        s3_archive_metadata: This dict will be used as the S3 metadata in the file in archive destination. Metadata in
+                             the source file will be replaced.
+
+        """
+        source_bucket = self.connection_config.get('s3_bucket')
+
+        # Get archive s3_bucket from config, or use same bucket if not specified
+        archive_bucket = self.connection_config.get('archive_load_files_s3_bucket', source_bucket)
+
+        # Determine prefix to use in archive s3 bucket
+        default_archive_prefix = 'archive'
+        archive_prefix = self.connection_config.get('archive_load_files_s3_prefix', default_archive_prefix)
+        prefixed_archive_key = '{}/{}'.format(archive_prefix, s3_archive_key)
+
+        copy_source = '{}/{}'.format(source_bucket, s3_source_key)
+
+        self.logger.info('Copying %s to archive location %s', copy_source, prefixed_archive_key)
+        self.upload_client.copy_object(copy_source, archive_bucket, prefixed_archive_key, s3_archive_metadata)
```

Comment:
how does the authentication works if it's two different buckets?
Can we use the same s3 upload_client for both bucket the we initialised at https://github.com/transferwise/pipelinewise-target-snowflake/blob/4de4250b025713c1ee2bf726600da8b53d204473/target_snowflake/upload_clients/s3_upload_client.py#L21 ?

Reply:
From [Boto3 docs](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.copy_object):
>All copy requests must be authenticated. Additionally, you must have read access to the source object and write access to the destination bucket. For more information, see REST Authentication . Both the Region that you want to copy the object from and the Region that you want to copy the object to must be enabled for your account.

So, if we want to use the `copy_object` method (rather than uploading object to two different buckets and using two sets of credentials), the credentials used for Snowflake imports must be valid for writing to the archive bucket as well. I tested this in staging and it works. Not sure how different region or endpoint url would play out.